File: appcore/src/main/java/mobi/cangol/mobile/service/route/RouteBuilder.java
package mobi.cangol.mobile.service.route;

import android.content.Context;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.SparseArray;

import java.io.Serializable;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import mobi.cangol.mobile.utils.UrlUtils;

/**
 * Created by xuewu.wei on 2018/10/16.
 */
public class RouteBuilder {
    private String path = null;
    private Bundle bundle = null;
    private RouteServiceImpl routeService;

    protected RouteBuilder(RouteServiceImpl routeService, String path) {
        this.routeService = routeService;
        this.bundle = new Bundle();
        this.path = path;
        if(path.contains("?")){
            String query =path.substring(path.indexOf("?")+1);
            for (final String entry : query.split("&")) {
                final String[] keyValue = entry.split("=");
                if (keyValue.length != 2) {
                    continue;
                }
                bundle.putString(keyValue[0], keyValue[1]);
            }
        }
    }

    protected String getPath() {
        return path;
    }

    protected Bundle getBundle() {
        return bundle;
    }

    public void navigation(Context context) {
        this.navigation(context,false);
    }

    public void navigation(Context context,boolean newStack) {
        this.routeService.handleNavigation(getPath(),getBundle(),context,newStack);
    }

    public RouteBuilder putString(String key, String value) {
        this.bundle.putString(key, value);
        return this;
    }

    public RouteBuilder putParcelable(String key, Parcelable value) {
        this.bundle.putParcelable(key, value);
        return this;
    }

    public RouteBuilder putChar(String key, char value) {
        this.bundle.putChar(key, value);
        return this;
    }

    public RouteBuilder putFloat(String key, float value) {
        this.bundle.putFloat(key, value);
        return this;
    }

    public RouteBuilder putShort(String key, short value) {
        this.bundle.putShort(key, value);
        return this;
    }

    public RouteBuilder putDouble(String key, Double value) {
        this.bundle.putDouble(key, value);
        return this;
    }

    public RouteBuilder putInt(String key, int value) {
        this.bundle.putInt(key, value);
        return this;
    }

    public RouteBuilder putSerializable(String key, Serializable value) {
        this.bundle.putSerializable(key, value);
        return this;
    }

    public RouteBuilder putLong(String key, long value) {
        this.bundle.putLong(key, value);
        return this;
    }

    public RouteBuilder putAll(Bundle bundle) {
        this.bundle.putAll(bundle);
        return this;
    }

    public RouteBuilder putByte(String key, byte value) {
        this.bundle.putByte(key, value);
        return this;
    }

    public RouteBuilder putBoolean(String key, boolean value) {
        this.bundle.putBoolean(key, value);
        return this;
    }

    public RouteBuilder putBundle(String key, Bundle value) {
        this.bundle.putBundle(key, value);
        return this;
    }


    public RouteBuilder putIntArray(String key, int[] value) {
        this.bundle.putIntArray(key, value);
        return this;
    }


    public RouteBuilder putFloatArray(String key, float[] value) {
        this.bundle.putFloatArray(key, value);
        return this;
    }

    public RouteBuilder putCharArray(String key, char[] value) {
        this.bundle.putCharArray(key, value);
        return this;
    }


    public RouteBuilder putLongArray(String key, long[] value) {
        this.bundle.putLongArray(key, value);
        return this;
    }


    public RouteBuilder putDoubleArray(String key, double[] value) {
        this.bundle.putDoubleArray(key, value);
        return this;
    }

    public RouteBuilder putByteArray(String key, byte[] value) {
        this.bundle.putByteArray(key, value);
        return this;
    }

    public RouteBuilder putBooleanArray(String key, boolean[] value) {
        this.bundle.putBooleanArray(key, value);
        return this;
    }

    public RouteBuilder putStringArray(String key, String[] value) {
        this.bundle.putStringArray(key, value);
        return this;
    }

    public RouteBuilder putCharSequence(String key, CharSequence value) {
        this.bundle.putCharSequence(key, value);
        return this;
    }

    public RouteBuilder putCharSequenceArray(String key, CharSequence[] value) {
        this.bundle.putCharSequenceArray(key, value);
        return this;
    }

    public RouteBuilder putStringArrayList(String key, ArrayList<String> value) {
        this.bundle.putStringArrayList(key, value);
        return this;
    }


    public RouteBuilder putParcelableArrayList(String key, ArrayList<? extends Parcelable> value) {
        this.bundle.putParcelableArrayList(key, value);
        return this;
    }

    public RouteBuilder putIntegerArrayList(String key, ArrayList<Integer> value) {
        this.bundle.putIntegerArrayList(key, value);
        return this;
    }

    public RouteBuilder putSparseParcelableArray(String key, SparseArray<? extends Parcelable> value) {
        this.bundle.putSparseParcelableArray(key, value);
        return this;
    }
}

File: appcore/src/main/java/mobi/cangol/mobile/service/route/RouteBuilder.kt
package mobi.cangol.mobile.service.route

import android.content.Context
import android.os.Bundle
import android.os.Parcelable
import android.util.SparseArray
import java.io.Serializable
import java.util.*


/**
 * Created by xuewu.wei on 2018/10/16.
 */
open class RouteBuilder(private val routeService: RouteServiceImpl, path: String) {
    private var path: String? = null
    private val bundle = Bundle()
    private var context: Context? = null

    init {
        this.path = path
    }

    fun getPath(): String {
        return this.path!!
    }

    fun getBundle(): Bundle {
        return bundle
    }

    fun getContext(): Context {
        return this.context!!
    }

    fun navigation(context: Context) {
        this.context = context
        val clazz = routeService.getClassByPath(path!!)
        if (clazz != null) {
            routeService.handleNavigation(clazz, bundle, context)
        }
    }

    fun putString(key: String, value: String): RouteBuilder {
        this.bundle.putString(key, value)
        return this
    }

    fun putParcelable(key: String, value: Parcelable): RouteBuilder {
        this.bundle.putParcelable(key, value)
        return this
    }

    fun putChar(key: String, value: Char): RouteBuilder {
        this.bundle.putChar(key, value)
        return this
    }

    fun putFloat(key: String, value: Float): RouteBuilder {
        this.bundle.putFloat(key, value)
        return this
    }

    fun putShort(key: String, value: Short): RouteBuilder {
        this.bundle.putShort(key, value)
        return this
    }

    fun putDouble(key: String, value: Double): RouteBuilder {
        this.bundle.putDouble(key, value)
        return this
    }

    fun putInt(key: String, value: Int): RouteBuilder {
        this.bundle.putInt(key, value)
        return this
    }

    fun putSerializable(key: String, value: Serializable): RouteBuilder {
        this.bundle.putSerializable(key, value)
        return this
    }

    fun putLong(key: String, value: Long): RouteBuilder {
        this.bundle.putLong(key, value)
        return this
    }

    fun putAll(bundle: Bundle): RouteBuilder {
        this.bundle.putAll(bundle)
        return this
    }

    fun putByte(key: String, value: Byte): RouteBuilder {
        this.bundle.putByte(key, value)
        return this
    }

    fun putBoolean(key: String, value: Boolean): RouteBuilder {
        this.bundle.putBoolean(key, value)
        return this
    }

    fun putBundle(key: String, value: Bundle): RouteBuilder {
        this.bundle.putBundle(key, value)
        return this
    }


    fun putIntArray(key: String, value: IntArray): RouteBuilder {
        this.bundle.putIntArray(key, value)
        return this
    }


    fun putFloatArray(key: String, value: FloatArray): RouteBuilder {
        this.bundle.putFloatArray(key, value)
        return this
    }

    fun putCharArray(key: String, value: CharArray): RouteBuilder {
        this.bundle.putCharArray(key, value)
        return this
    }


    fun putLongArray(key: String, value: LongArray): RouteBuilder {
        this.bundle.putLongArray(key, value)
        return this
    }


    fun putDoubleArray(key: String, value: DoubleArray): RouteBuilder {
        this.bundle.putDoubleArray(key, value)
        return this
    }

    fun putByteArray(key: String, value: ByteArray): RouteBuilder {
        this.bundle.putByteArray(key, value)
        return this
    }

    fun putBooleanArray(key: String, value: BooleanArray): RouteBuilder {
        this.bundle.putBooleanArray(key, value)
        return this
    }

    fun putStringArray(key: String, value: Array<String>): RouteBuilder {
        this.bundle.putStringArray(key, value)
        return this
    }

    fun putCharSequence(key: String, value: CharSequence): RouteBuilder {
        this.bundle.putCharSequence(key, value)
        return this
    }

    fun putCharSequenceArray(key: String, value: Array<CharSequence>): RouteBuilder {
        this.bundle.putCharSequenceArray(key, value)
        return this
    }

    fun putStringArrayList(key: String, value: ArrayList<String>): RouteBuilder {
        this.bundle.putStringArrayList(key, value)
        return this
    }


    fun putParcelableArrayList(key: String, value: ArrayList<out Parcelable>): RouteBuilder {
        this.bundle.putParcelableArrayList(key, value)
        return this
    }

    fun putIntegerArrayList(key: String, value: ArrayList<Int>): RouteBuilder {
        this.bundle.putIntegerArrayList(key, value)
        return this
    }

    fun putSparseParcelableArray(key: String, value: SparseArray<out Parcelable>): RouteBuilder {
        this.bundle.putSparseParcelableArray(key, value)
        return this
    }
}
